tasks/insights-client.yml

```diff
@@ -133,6 +223,46 @@
           or "Registered" in __rhc_insights_status.stdout
       changed_when: true
 
+- name: Configure file redaction
+  when: (rhc_insights.file_redaction.commands is defined) or
```

Comment:
Does there need to be a check first for `rhc_insights.file_redaction is defined`?

Reply:
AFAICS `rhc_insights.file_redaction` is defined in `defaults/main.yml`. As `defaults/main.yml` as well as `tasks/insights-client.yml` are part of one role and shipped together I do not see a scenario where `rhc_insights.file_redaction` would not be defined. So with my current knowledge and understanding I would say there is no need to check first for `rhc_insights.file_redaction is defined`.

I see that most if not all tasks in `tasks/insights-client.yml` check whether variables are defined although all of these variables are defined in `defaults/main.yml`. TBH I do not understand why these checks are necessary. Are they in place for "just in case" scenarios or is there some reason I do not see?

Following my own argument I would suggest removing all the `var is defined` checks where the var is defined in `defaults/main.yml` as the check is not striclty necessary. But I might miss a good reason here.

Reply:
> AFAICS `rhc_insights.file_redaction` is defined in `defaults/main.yml`.

That unfortunately doesn't mean anything.

> As `defaults/main.yml` as well as `tasks/insights-client.yml` are part of one role and shipped together I do not see a scenario where `rhc_insights.file_redaction` would not be defined.

What if I run the role like this?
```
inventory:....
  rhc_insights:
    foo: bar
```
Defining a host variable takes precedence over settings in defaults/main.yml.  The same applies to setting the variable at the play level or as one of the `vars` in the `include_role` call.  Setting `rhc_insights` in this way *completely replaces the existing contents defined in defaults/main.yml*.  https://docs.ansible.com/ansible/latest/reference_appendices/general_precedence.html#variables

So - while "top level" variables like `rhc_insights` will always be defined, their "sub-parameters" may or may not be defined - that is - `rhc_insights is defined` will always return `true`, but `rhc_insights.file_redaction is defined` can return `true` or `false`.

If you don't believe me, try running the role with different values and see what happens.

> So with my current knowledge and understanding I would say there is no need to check first for `rhc_insights.file_redaction is defined`.
> 
> I see that most if not all tasks in `tasks/insights-client.yml` check whether variables are defined although all of these variables are defined in `defaults/main.yml`. TBH I do not understand why these checks are necessary. Are they in place for "just in case" scenarios or is there some reason I do not see?

Yes, and if some checks are missing, they should be added in another PR.

TBH, I'm really not sure why "sub-parameter" keys and values are defined in `defaults/main.yml`, except possibly as a convenience when the user does not provide a value for `rhc_insights` at all.  There is no mechanism in Ansible like "merge the default dict valued parameter `rhc_insights` defined in `defaults/main.yml` with the `rhc_insights` parameter given by the user".  If the user defines `rhc_insights`, then the value in `defaults/main.yml` is completely replaced.

> 
> Following my own argument I would suggest removing all the `var is defined` checks where the var is defined in `defaults/main.yml` as the check is not striclty necessary. But I might miss a good reason here.